R/learner_BART_surv_bart.R

```diff
@@ -0,0 +1,212 @@
+          openmp = p_lgl(default = TRUE, tags = "predict"),
+          quiet = p_lgl(default = TRUE, tags = "predict"),
+          importance = p_fct(default = "count", levels = c("count", "prob"), tags = "train"),
+          which.curve = p_dbl(special_vals = list("mean"), tags = "predict")
+        )
+
+        # custom defaults
+        param_set$values = list(mc.cores = 1, quiet = TRUE, importance = "count",
+          which.curve = 0.5) # 0.5 quantile => median posterior
+
+        super$initialize(
+          id = "surv.bart",
+          packages = "BART",
+          feature_types = c("logical", "integer", "numeric"),
+          predict_types = c("crank", "distr"),
+          param_set = param_set,
+          properties = c("importance", "missings"),
+          man = "mlr3extralearners::mlr_learners_surv.bart",
+          label = "Bayesian Additive Regression Trees"
+        )
+      },
+
+      #' @description
+      #' Two types of importance scores are supported based on the initial value
+      #' for the parameter `importance`:
+      #' 1. `prob`: The mean selection probability of each feature in the trees,
+      #' extracted from the slot `varprob.mean`.
+      #' If `sparse = FALSE` (default), this is a fixed constant.
+      #' Recommended to use this option when `sparse = TRUE`.
+      #' 2. `count`: The observed count of each feature in the trees, extracted
+      #' from the slot `varcount.mean`.
+      #' This is the default importance scores.
+      #'
+      #' In both cases, higher values signify more important variables.
+      #'
+      #' @return Named `numeric()`.
+      importance = function() {
+        if (is.null(self$model$model)) {
+          stopf("No model stored")
+        }
+
+        pars = self$param_set$get_values(tags = 'train')
+
+        if (pars$importance == 'prob') {
+          sort(self$model$model$varprob.mean[-1], decreasing = T)
+        } else {
+          sort(self$model$model$varcount.mean[-1], decreasing = T)
+        }
+      }
+    ),
+
+    private = list(
+      .train = function(task) {
+        pars = self$param_set$get_values(tags = "train")
+        pars$importance = NULL # not used in the train function
+
+        x.train = as.data.frame(task$data(cols = task$feature_names))
+        times   = task$truth()[,1]
+        delta   = task$truth()[,2] # delta => status
+
+        list(
+          model = invoke(
+            BART::mc.surv.bart,
+            x.train = x.train,
+            times = times,
+            delta = delta,
+            .args = pars
+          ),
+          # need these for predict
+          x.train = x.train,
+          times = times,
+          delta = delta
+        )
+      },
+
+      .predict = function(task) {
+        # get parameters with tag "predict"
+        pars = self$param_set$get_values(tags = "predict")
+
+        # get newdata and ensure same ordering in train and predict
+        x.test = as.data.frame(ordered_features(task, self))
```

Comment:
is the data.frame conversion needed?

Reply:
Sadly yes, otherwise you get the error:
```
Error in bartModelMatrix(x.train, numcut, usequants = usequants, cont = cont,  : 
  Expecting either a factor, a vector, a matrix or a data.frame
```